### PR TITLE
Raami

### PR DESCRIPTION
**Pull request by lebanonsky**
ready for production

### CODE REVIEW

**Comment by lebanonsky** [comment]
afaik have not made any changes to kansa/admin

File: API.md [hunk 1]
```diff
@@ -38,6 +38,8 @@ some relevant data and/or a `message` field.
   * [`GET /api/hugo/admin/ballots/:category`](#get-apihugoadminballotscategory)
   * [`GET /api/hugo/admin/canon`](#get-apihugoadmincanon)
   * [`POST /api/hugo/admin/canon/:id`](#post-apihugoadmincanonid)
+  * [WebSocket: `wss://server/api/hugo/admin/canon-updates`](#websocket-wssserverapihugoadmincanonupdates)
+* [Raami exhibitions](#raami)
```
Comment (eemeli) [inline API.md]: Should be at the end, not in the middle of the hugo-admin links. Also, the title link should be to `#raami-exhibition` and the individual endpoints included here.

File: API.md [hunk 2]
```diff
@@ -575,3 +577,146 @@ events to signal `'Unauthorized'` and `'Not Found'` (respectively) to the client
...
+- Parameters: `id` (required)
+
+Full details for singular artist.
+
+#### Response
+```
+  { continent: 'Europe',
+    url: 'http://www.example.com',
+    filename: 'portfolio.pdf',
+    portfolio: [bytes],
```
Comment (eemeli) [inline API.md]: This at least should be expanded. As is, it's a mystery.

File: API.md [hunk 3]
```diff
@@ -575,3 +577,146 @@ events to signal `'Unauthorized'` and `'Not Found'` (respectively) to the client
...
+    category: 'Cover artist',
+    orientation: 'Fantasy',
+    description: 'Who is the artist etc.',
+    trasnport: 'Air mail'
```
Comment (eemeli) [inline API.md]: `trasnport`?

File: API.md [hunk 4]
```diff
@@ -575,3 +577,146 @@ events to signal `'Unauthorized'` and `'Not Found'` (respectively) to the client
...
+- Parameters: `id` (required) artists id
+
+Ids of works for particular artists
+
+#### Response
+```
+[
+  { id: 1 },
+  …
+]
```
Comment (eemeli) [inline API.md]: Why an array of `{ id }` objects, rather than just an array of ids?

File: API.md [hunk 5]
```diff
@@ -575,3 +577,146 @@ events to signal `'Unauthorized'` and `'Not Found'` (respectively) to the client
...
+- Parameters: `person_id`, `continent`, `url`, `filename`, `portfolio`, `category`, `orientation`, `description`, `transport`
+
+Insert new artist's details.
+
+#### Response
+```
+{
+  status: 'success',
+  inserted: [ 'id', 'continent', 'url','filename', 'portfolio', 'category', 'orientation', 'description', 'transport' ]
+}
```
Comment (eemeli) [inline API.md]: According to the implementation, `inserted` is in fact the id of the new artist.

File: API.md [hunk 6]
```diff
@@ -575,3 +577,146 @@ events to signal `'Unauthorized'` and `'Not Found'` (respectively) to the client
...
+```
+
+### `PUT /api/raami/artist/:id`
+
+- Parameters: `id` (required), `continent`, `url`, `filename`, `portfolio`, `category`, `orientation`, `description`, `transport`
+
+Update new artist's details.
```
Comment (eemeli) [inline API.md]: It should be made clear what happens if some fields are missing.

File: API.md (hunk 4 above)
Comment (eemeli) [inline API.md]: On further thought, why doesn't this endpoint provide directly all of the works' data, rather than just their ids?

File: API.md [hunk 7]
```diff
@@ -575,3 +577,146 @@ events to signal `'Unauthorized'` and `'Not Found'` (respectively) to the client
...
+#### Response
+```
+[
+  { artist_id: 1,
+    title: 'Book cover',
+    width: 10.0,
+    height: 10.0,
+    technique: 'Oil on canvas',
+    graduation: 0,
+    filename: 'file.jpg',
```
Comment (eemeli) [inline API.md]: What is `filename` used for?

File: API.md [hunk 8]
```diff
@@ -575,3 +577,146 @@ events to signal `'Unauthorized'` and `'Not Found'` (respectively) to the client
...
+    image: bytes,
+    price: 100.0
```
Comment (eemeli) [inline API.md]: Rather than using floating-point values, prices should be in integer cents.

File: API.md [hunk 9]
```diff
@@ -575,3 +577,146 @@ events to signal `'Unauthorized'` and `'Not Found'` (respectively) to the client
...
+
+Full details for singular work.
+
+#### Response
+```
+[
+  { artist_id: 1,
+    title: 'Book cover',
+    width: 10.0,
+    height: 10.0,
```
Comment (eemeli) [inline API.md]: Would rather prefer integer millimeters.

File: API.md [hunk 10]
```diff
@@ -575,3 +577,146 @@ events to signal `'Unauthorized'` and `'Not Found'` (respectively) to the client
...
+  status: 'success',
+  inserted: [ 'artist_id', 'title', 'width', 'height', 'technique', 'graduation', 'filename', 'image', 'price' ]
+}
+```
+
+### `PUT /api/raami/work/:id`
+
+- Parameters: `id`, `artist_id`, `title`, `width`, `height`, `technique`, `graduation`, `filename`, `image`, `price`
+
+Update works's details.
```
Comment (eemeli) [inline API.md]: Same as with the other `PUT`, should say what happens if parameters are missing or `undefined`.

File: kansa/admin/package.json [hunk 11]
```diff
@@ -12,7 +12,7 @@
   "scripts": {
     "build": "webpack -d",
     "build:prod": "NODE_ENV=production webpack -p",
-    "start": "webpack-dev-server -d",
+    "start": "webpack-dev-server -d --hot --inline",
```
Comment (eemeli) [inline kansa/admin/package.json]: Why is this here?

File: nginx/index.html [hunk 12]
```diff
@@ -6,5 +6,5 @@
 </head>
 <body>
 <section id="react"></section>
-<script type="text/javascript" src="https://worldcon75.github.io/client/master/bundle.js"></script></body>
+<script type="text/javascript" src="https://worldcon75.github.io/client/raami/bundle.js"></script></body>
```
Comment (eemeli) [inline nginx/index.html]: This will need to get removed in the merge.

File: postgres/init/40-raami-init.sql [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    RETURN null;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE FUNCTION work_notify() RETURNS trigger as $$
+BEGIN
+    PERFORM pg_notify('work', row_to_json(NEW)::text);
+    RETURN null;
+END;
+$$ LANGUAGE plpgsql;
```
Comment (eemeli) [inline postgres/init/40-raami-init.sql]: Are these functions used anywhere? If not, they should be removed.

File: raami/package.json [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+    "body-parser": "^1.13.3",
+    "connect-pg-simple": "^3.1.2",
+    "cookie-parser": "~1.3.5",
+    "debug": "~2.2.0",
+    "express": "~4.13.1",
+    "express-session": "^1.15.0",
+    "morgan": "~1.6.1",
+    "pg-promise": "^3.2.3"
+  }
+}
```
Comment (eemeli) [inline raami/package.json]: Should include `"license": "Apache-2.0"` and a non-zero version number.

File: raami/queries.js [hunk 15]
```diff
@@ -0,0 +1,269 @@
+// var promise = require("bluebird");
+
+// var options = {
+//   // Initialization Options
+//   promiseLib: promise
+// };
+
+// var pgp = require("pg-promise")(options);
+// var db = pgp(process.env.DATABASE_URL)
```
Comment (eemeli) [inline raami/queries.js]: Remove these comments.

File: raami/queries.js [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+      FROM Artist
+     WHERE person_id = $1`, _id
+  )
+    .then(function (data) {
+      res.status(200)
+        .json(data);
+    })
+    .catch(function (err) {
+      return next(err);
+    });
```
Comment (eemeli) [inline raami/queries.js]: Rather, just use `.catch(next);`

File: raami/queries.js [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+                 )
+          VALUES (
+                   $(person_id), $(name), $(continent), $(url), $(filename),
+                   $(filedata), $(description), $(transport), $(legal),
+                   ($(auction)), ($(print)), $(digital), $(agent), $(contact), $(waitlist), ($(postage))
+                 )
+       RETURNING id`, req.body
+  )
+    .then(function (data) {
+      console.log(data.id)
```
Comment (eemeli) [inline raami/queries.js]: Remove

File: raami/queries.js [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+  req.app.locals.db.one(`
+     INSERT INTO Artist
+                 (
+                   person_id, name, continent, url, filename, filedata,
+                   description, transport, legal, auction, print, digital, agent, contact, waitlist, postage
+                 )
+          VALUES (
+                   $(person_id), $(name), $(continent), $(url), $(filename),
+                   $(filedata), $(description), $(transport), $(legal),
+                   ($(auction)), ($(print)), $(digital), $(agent), $(contact), $(waitlist), ($(postage))
```
Comment (eemeli) [inline raami/queries.js]: Why the extra parentheses?

File: raami/queries.js [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+      req.body.legal,
+      (req.body.auction),
+      (req.body.print),
+      req.body.digital,
+      req.body.agent,
+      req.body.contact,
+      req.body.waitlist,
+      (req.body.postage),
+      _id
+    ]
```
Comment (eemeli) [inline raami/queries.js]: Again, extra parentheses.

Should require `raami_admin` rights, or check that the email address of `person_id` matches `req.session.user.email`.

Also, would be clearer with an object rather than a 16-value array. Take a look at the pg-promise helpers [`sets`](http://vitaly-t.github.io/pg-promise/helpers.html#.sets) and [`update`](http://vitaly-t.github.io/pg-promise/helpers.html#.update).

File: raami/queries.js [hunk 20]
```diff
@@ -0,0 +1,269 @@
...
+    .then(function () {
+      res.status(200)
+        .json({
+          status: 'success',
+          updated: req.body
+        });
+    })
+    .catch(function (err) {
+      return next(err);
+    });
```
Comment (eemeli) [inline raami/queries.js]: `.catch(next)`

File: raami/queries.js [hunk 21]
```diff
@@ -0,0 +1,269 @@
...
+  )
+    .then(function (data) {
+      res.status(200)
+        .json({
+          works: data,
+        });
+    })
+    .catch(function (err) {
+      return next(err);
+    });
```
Comment (eemeli) [inline raami/queries.js]: `.catch(next)`

File: raami/queries.js [hunk 22]
```diff
@@ -0,0 +1,269 @@
...
+    .then(function (data) {
+      res.status(200)
+        .json({
+          status: 'success',
+          insrted: data.id
+        });
+    })
+    .catch(function (err) {
+      return next(err);
+    });
```
Comment (eemeli) [inline raami/queries.js]: `.catch(next)`

File: raami/queries.js [hunk 23]
```diff
@@ -0,0 +1,269 @@
...
+      req.body.depth,
+      req.body.gallery,
+      req.body.filename,
+      req.body.filedata,
+      req.body.price,
+      req.body.year,
+      req.body.orientation,
+      req.body.technique,
+      _id
+    ]
```
Comment (eemeli) [inline raami/queries.js]: See [above](#r100045779).

File: raami/routes.js [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+  res.header("Access-Control-Allow-Credentials",'true')
+  next();
+ });
+
+function authenticate(req, res, next) {
+  if (req.session && req.session.user && req.session.user.email) next();
+  else res.status(401).json({ status: 'unauthorized' });
+}
+
+router.use(authenticate);
```
Comment (eemeli) [inline raami/routes.js]: This will only verify that the user is signed in; it does not verify whether the user should have any access to this specific artist's information. Also, we should add a `raami_admin` entry [here](/worldcon75/api/blob/master/postgres/init/10-admin-init.sql#L2) and [here](/worldcon75/api/blob/master/kansa/server/lib/user.js#L39) that grants wider access.

File: raami/queries.js [hunk 25]
```diff
@@ -0,0 +1,269 @@
...
+};
+
+function getPeople(req, res, next) {
+  var _id = (req.params.id);                             
+  req.app.locals.db.any(`
+    SELECT id, person_id, name, continent, url, filename, filedata, description,
+           transport, legal, auction, print, digital, agent, contact, waitlist, postage
+      FROM Artist
+     WHERE person_id = $1`, _id
+  )
```
Comment (eemeli) [inline raami/queries.js]: Should require `raami_admin` rights, or check that the email address of `person_id` matches `req.session.user.email`.

File: raami/queries.js [hunk 26]
```diff
@@ -0,0 +1,269 @@
...
+    });
+}
+
+function getArtist(req, res, next) {
+  var _id = (req.params.id);
+  req.app.locals.db.one(`
+    SELECT *
+      FROM Artist
+     WHERE id = $1`, _id
+  )
```
Comment (eemeli) [inline raami/queries.js]: Should require `raami_admin` rights, or check that the email address of `person_id` matches `req.session.user.email`.

File: raami/queries.js [hunk 27]
```diff
@@ -0,0 +1,269 @@
...
+                   person_id, name, continent, url, filename, filedata,
+                   description, transport, legal, auction, print, digital, agent, contact, waitlist, postage
+                 )
+          VALUES (
+                   $(person_id), $(name), $(continent), $(url), $(filename),
+                   $(filedata), $(description), $(transport), $(legal),
+                   ($(auction)), ($(print)), $(digital), $(agent), $(contact), $(waitlist), ($(postage))
+                 )
+       RETURNING id`, req.body
+  )
```
Comment (eemeli) [inline raami/queries.js]: Should require `raami_admin` rights, or check that the email address of `person_id` matches `req.session.user.email`.

File: raami/queries.js [hunk 28]
```diff
@@ -0,0 +1,269 @@
...
+/**** WORKS ***/
+
+function getWorks(req, res, next) {
+  var _id = (req.params.id);
+  req.app.locals.db.any(`
+    SELECT id, artist_id, title, width, height, depth, gallery, orientation, technique,
+           filename, filedata, year, price
+      FROM Works
+     WHERE artist_id = $1`, _id
+  )
```
Comment (eemeli) [inline raami/queries.js]: Should require `raami_admin` rights, or check that the email address of `person_id` matches `req.session.user.email`.

File: raami/queries.js [hunk 29]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+function getWork(req, res, next) {
+  var _id = (req.params.id);
+  req.app.locals.db.one(`
+    SELECT artist_id, title, width, height, depth, gallery, orientation, technique,
+           filename, filedata, year, price
+      FROM Works
+     WHERE id = $1`, _id
+  )
```
Comment (eemeli) [inline raami/queries.js]: Should require `raami_admin` rights, or check that the email address of `person_id` matches `req.session.user.email`.

File: raami/queries.js [hunk 30]
```diff
@@ -0,0 +1,269 @@
...
+                  artist_id, title, width, height, depth, gallery, orientation,
+                  technique, filename, filedata, year, price
+                )
+         VALUES (
+                  $(artist_id), $(title), $(width), $(height), $(depth), $(gallery),
+                  $(orientation), $(technique), $(filename), $(filedata),
+                  $(year), $(price)
+                )
+        RETURNING id`, req.body
+  )
```
Comment (eemeli) [inline raami/queries.js]: Should require `raami_admin` rights, or check that the email address of `person_id` matches `req.session.user.email`.

File: raami/queries.js [hunk 31]
```diff
@@ -0,0 +1,269 @@
...
+        });
+    })
+    .catch(function (err) {
+      return next(err);
+    });
+}
+
+function removeWork(req, res, next) {
+  var _id = (req.params.id);
+  req.app.locals.db.result('DELETE FROM Works WHERE id = $1', _id)
```
Comment (eemeli) [inline raami/queries.js]: Should require `raami_admin` rights, or check that the email address of `person_id` matches `req.session.user.email`.

File: API.md [hunk 32]
```diff
@@ -575,3 +577,146 @@ events to signal `'Unauthorized'` and `'Not Found'` (respectively) to the client
...
+    image: bytes,
+    price: 100.0
+    }
+  …
+]
+```
+
+### `POST /api/raami/work`
+
+- Parameters: `artist_id`, `title`, `width`, `height`, `technique`, `graduation`, `filename`, `image`, `price`
```
Comment (eemeli) [inline API.md]: Missing `depth`.
